### DISCUSSION TITLE
Server actions are now broken on vercel (preview and prod) - 405 Method Not Allowed error

### DISCUSSION THREAD

**Discussion by Xexr** [vercel/community#6651] [Help] [answered]
### Summary

What ever updates were made to the vercel build process or runtime over the last ~12 hours appear to have broken server actions on vercel.

I have a deployment from yesterday that works perfectly.

All new deployments (including the exact same code/commit) now exhibit "405 Method Not Allowed error" on all server action calls.

The same code continues to work in the original deployment and locally.

Limited to the edge runtime

### Steps to Reproduce

Deploy any server actions to vercel using the edge runtime

**Comment by vyrnix9582046173** [comment]
also ran into this, it seems specific to `edge` runtime specifically for me, might be a cloudflare issue?

edit: noticed this about 7 hours ago initially

**Comment by Xexr** [comment]
Thanks Chris - confirmed that its edge specific for me too. Disabling edge and running on serverless doesn't exhibit the behaviour.

I'm based in the UK, but don't think its geo limited as tried accessing via VPN to US and got the same error.

I noticed it around 11.30pm last night  (c. 12 hours ago)

**Comment by Xexr** [comment]
Support got back to me, looks like its a vercel cli issue. In the interim, you can force the old version of the CLI by setting the following as an environment variable and redeploying:

VERCEL_CLI_VERSION=vercel@33.7.1

**Reply by aleksa-codes** [reply]
In my case a redirect is not working properly inside a server action, and also an API route that is not using edge runtime returns 405. This fixed it for now.

**Reply by alexdaley** [reply]
Did support give you any insight as to the root cause or an expected remediation beyond overriding the CLI version?

**Reply by Xexr** [reply]
No they didn't go into detail, just that they thought it was caused by a recent CLI release.

I no longer get the error, so it would appear they've fixed the root cause (or at least rolled back vercel by default)

**Reply by brhx** [reply]
It appears they've rolled it back in all builds. My new builds use `33.7.1` without the env var override set.

**Comment by callmefredcom** [comment]
I tried the fix VERCEL_CLI_VERSION=vercel@33.7.1.

It solved the 405 but I get a 504 (timeout) when trying to upload a basic (not heavy) image to Blob Storage via my API endpoint.

This thing is driving me crazy, to say the least...

Here's the code of my API endpoint. What's wrong with it?

`import { put } from '@vercel/blob';
import type { NextApiResponse, NextApiRequest, PageConfig } from 'next';

export default async function handler(
  request: NextApiRequest,
  response: NextApiResponse,
) {
  // Only allow POST requests, explicitly reject other methods
  if (request.method !== 'POST') {
    console.error(`Attempted to handle ${request.method} method`);
    return response.status(405).end('Method Not Allowed');
  }

  try {
    const blob = await put(request.query.filename as string, request, {
      access: 'public',
    });
    return response.status(200).json(blob);
  } catch (error) {
    console.error('Error processing request', error);
    return response.status(500).json({ error: 'Internal Server Error' });
  }
}
`

**Reply by aleksa-codes** [reply]
Good thing to hear I am not alone on this one. I think I am losing the little hair I have left working with juggling between server and client components, not knowing when to use edge runtime, server actions, and dealing with the Vercel CLI bug/issue. Sometimes I think I should have stuck with React + Node and called it a day.

One thing I can suggest regarding your problem: the first thing I always do when I get a timeout is to add:
`export const runtime = 'edge';` before your exported function because of what I just mentioned. There is some timeout limit on the free Vercel tier, or something like that. Another thing, I can't understand if this is an API route and are you using App router or Page router? If you are using the App router, this is the way of writing API routes now: [Next.js Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers)

**Reply by callmefredcom** [reply]
UPDATE: I added export const runtime = 'edge'; to my API endpoint code and the file is now uploaded.
But... it's not properly formatted, it's corrupted...
I need to debug that.

```
import { put } from '@vercel/blob';

export const runtime = 'edge';

export default async function handler(request: Request) {
  // Only allow POST requests, explicitly reject other methods
  if (request.method !== 'POST') {
    console.error(`Attempted to handle ${request.method} method`);
    return new Response('Method Not Allowed', { status: 405 });
  }

  try {
    const url = new URL(request.url);
    const filename = url.searchParams.get('filename');
    if (!filename) {
      return new Response('Bad Request: filename is missing', { status: 400 });
    }
    const body = await request.arrayBuffer();
    const blob = await put(filename, body, {
      access: 'public',
    });
    return new Response(JSON.stringify(blob), { status: 200 });
  } catch (error) {
    console.error('Error processing request', error);
    return new Response(JSON.stringify({ error: 'Internal Server Error' }), { status: 500 });
  }
}
```

**Reply by callmefredcom** [reply]
By the way, I found the solution here: https://github.com/vercel/examples/blob/main/storage/blob-starter/app/api/upload/route.ts
And then I saw your answer @aleksa-codes, which confirmed the fix ;) Thanks

**Reply by aleksa-codes** [reply]
@callmefredcom Following up on my comment, I made a mistake about the Vercel Hobby free tier.

You can actually set the Node maximum function duration to up to 60 seconds on the free tier and avoid timeouts. Here's a [guide on avoiding timeouts](https://vercel.com/guides/what-can-i-do-about-vercel-serverless-functions-timing-out) and a [configuration guide](https://vercel.com/docs/functions/configuring-functions/duration#node.js-next.js-%3E=-13.5-or-higher-sveltekit-astro-nuxt-and-remix).

To configure this, add or replace `export const runtime = 'edge';` with `export const maxDuration = 20;` for example (default is 10s). Using this, I managed to move all my code from the Edge runtime to Node, adjusting `maxDuration` as needed. My apologies for the confusion.

Edit: for Next.js with Pages Router it is for example: `export const config = {
  maxDuration: 20
};`

**Reply by leerob** [reply]
Glad that worked for you! I also recommend that approach.

**Comment by callmefredcom** [comment]
Alleluia, this is working now: https://frontendtest-self.vercel.app/dynamic

So to fix the "empty / corrupted image" issue, I just followed the tips given by Vercel's AI Assistant.

Here's the full code, working fine:

CLIENT SIDE:

The file input: 

```
   <div style={{ display: 'flex', justifyContent: 'center' }}>
  <input type="file" accept="image/*" onChange={handleImageUpload} style={{ display: 'none' }} id="imageUpload" />
  <label htmlFor="imageUpload" style={{ display: 'flex', alignItems: 'center', justifyContent: 'center', cursor: 'pointer' }}>
    <AiFillPicture style={{ color: '#000', fontSize: '40px', margin: '20px' }} />
  </label>

```

The upload function

```
const handleImageUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
  if (event.target.files) {
    const file = event.target.files[0];

    try {
      const response = await fetch(`/api/upload?filename=${encodeURIComponent(file.name)}`, {
        method: 'POST',
        body: file, 
        headers: {
          'Accept': 'application/json', 
        },
      });

      if (!response.ok) {
        throw new Error('Failed to upload file');
      }

      const data = await response.json();
      setBgImage(data.url);
    } catch (error) {
      console.error('Error:', error);
      alert('Failed to upload image');
    }
  }
}

```

THE API ENDPOINT

```
import { put } from '@vercel/blob';

export const runtime = 'edge';

export default async function handler(request: Request) {
  // Only allow POST requests, explicitly reject other methods
  if (request.method !== 'POST') {
    console.error(`Attempted to handle ${request.method} method`);
    return new Response('Method Not Allowed', { status: 405 });
  }

  try {
    const url = new URL(request.url);
    const filename = url.searchParams.get('filename');
    if (!filename) {
      return new Response('Bad Request: filename is missing', { status: 400 });
    }
    const body = await request.arrayBuffer();
    const blob = await put(filename, body, {
      access: 'public',
    });
    return new Response(JSON.stringify(blob), { status: 200 });
  } catch (error) {
    console.error('Error processing request', error);
    return new Response(JSON.stringify({ error: 'Internal Server Error' }), { status: 500 });
  }
}
```

All good!
Thanks again @aleksa-codes for your help re: runtime

**Reply by aleksa-codes** [reply]
Nice, glad to hear it. Just one thing, as from the docs I linked to you and the solution you found, I think:
`export default async function handler(request: Request) {`
should be replaced with this if you are using App router Next.js:
`export async function POST(req: Request) {`

but maybe don't touch it, if it is working, idk 😂 just a hint

**Reply by callmefredcom** [reply]
I'm using the Page Router, so maybe the current version is ok.
I don't know actually :)

**Comment by callmefredcom** [comment]
The joy was short... I hit the 100Mb Outgoing Bandwidth Limit for Blob Storage in 1h... 
I'll refactor for S3.

**Reply by vvo** [reply]
Hey @callmefredcom, we just made two changes to offer a better Blob bandwidth package:
- Starting today, Hobby customers have 5GB of free monthly Vercel Blob bandwidth (before: 100 MB)
- Starting today, Pro and Pro trial customers have 10GB of free monthly Vercel Blob bandwidth (before: 1GB)

Hopefully, these new limits will provide better defaults so people can have a good experience while testing Vercel Blob.
Our previous limits (especially for Hobby customers) were too low, sorry about that! 👍.

**Comment by swarnava** [comment, marked as answer]
We have rolled out a feature flag now so you no longer need to pin CLI version anymore. We will incrementally test and rollout for everyone.

We recommend to remove the Environment Variable to receive future update released using CLI.

**Comment by kelihansen** [comment]
I still have to pin to `VERCEL_CLI_VERSION=vercel@33.7.1` to avoid 404s when redirecting from a server action as of this morning.

**Reply by swarnava** [reply]
We disabled the flag again. Can you go to Project > Security & Enable `Customer Success Code Visibility` temporarily? Please let us any one of the deployment link so we can track the deployment.

**Reply by swarnava** [reply]
We found the problem. Pushing a fix right now. In the meantime, we rolled back the change so you can remove the Environment variable again. :)

**Reply by aleksa-codes** [reply]
hi @swarnava, hate to be annoying 😅 but it is happening right now again

**Reply by ztanner** [reply]
Hi @aleksa-codes -- are you experiencing the issue in a build that you've pinned to an older CLI version, or on the latest?

**Reply by aleksa-codes** [reply]
hi @ztanner, on the latest version

**Reply by ztanner** [reply]
@aleksa-codes thanks for the confirmation -- are you able to share with me a build ID or deployment URL?

**Reply by aleksa-codes** [reply]
@ztanner I can't share the URL here since it is a private repo, here is the project ID from the general setting: `prj_8dswlkLWGkLcEhtc3bOTGHu1D44z`

Let me know if it is the wrong ID. Thanks

**Reply by ztanner** [reply]
Thanks @aleksa-codes -- could you also clarify the unexpected behavior you're seeing? Is there a runtime error, an unexpected status code, or something else? I'm not seeing any errors in the deployment/application logs.

**Reply by aleksa-codes** [reply]
@ztanner Thanks for the quick replies. It is related to the original comment here:
> I still have to pin to `VERCEL_CLI_VERSION=vercel@33.7.1` to avoid 404s when redirecting from a server action as of this morning.

In my case, I have a logout function as a server action that after a request is finished does `redirect(/);` and the redirect works, in the URL after the redirect I see the root/Homepage, but the page displays the 404 not-found page. I can even see the dynamic footer that I only have for the Homepage.

**Reply by ztanner** [reply]
I see! I'm going to temporarily remove you from the feature flag to investigate further. Thanks for letting us know!

**Comment by adilanchian** [comment]
we were also in the same boat today.

pinning `VERCEL_CLI_VERSION=vercel@33.7.1` got rid of our 405 error on our action call.

**Reply by swarnava** [reply]
We disabled the flag again. Can you go to Project > Security & Enable `Customer Success Code Visibility` temporarily? Please let us any one of the deployment link so we can track the deployment.

**Reply by swarnava** [reply]
We found the problem. Pushing a fix right now. In the meantime, we rolled back the change so you can remove the Environment variable again. :)

**Comment by brhx** [comment]
@swarnava This has started happening again on my deployments made today. On Vercel CLI 34.2.0

**Reply by ztanner** [reply]
Hi @sam3d -- can you share more information about the deploy that's not working / the behavior that you're seeing?

**Reply by darekrossman** [reply]
It started happening to us again too - any server action that passes through a URL with a rewrite in next.config returns a 404. The action still works, but it causes issues. 

We can also see that our build summary shows 82 functions where before it showed 44 - there are a bunch of new functions in there with `.action` appended. Builds without that worked.

**Reply by brhx** [reply]
I'm seeing a 405 (Method Not Allowed) but under the exact same conditions you've described, @darekrossman

**Reply by swarnava** [reply]
We are investigating this right now.

**Reply by swarnava** [reply]
While we investigating this, we have disabled the feature and a redeploy should see better results

**Reply by brhx** [reply]
This has now just started happening again. Seems to be mostly unrelated to the Vercel CLI version (except pinning back to 33.7.1 is the only reliable fix). How do I permanently opt-out of whatever feature is causing this issue?

**Reply by swarnava** [reply]
Hey @sam3d I just responded you over email!

**Reply by manu-r** [reply]
same issue. everything works fine in `production` while `preview` fails with 405. setting cli to 33 works, but had to downgrade to pnpm@8

**Reply by ztanner** [reply]
Hi @manu-r - would it be possible for you to share the impacted deployment ID? You can find this in the browser URL after clicking into the deployment in the Vercel dashboard.

**Reply by manu-r** [reply]
hi @ztanner,

I am guessing it's this one: `nscogg7rw`. I am not sure if you have access to it since it's a private repo.

**Reply by swarnava** [reply]
You can go to Vercel Dashboard > Project > specific deployment and share the link from your URL bar. Thanks

**Reply by manu-r** [reply]
hi @swarnava, don't want to share the url here as it's a private project. here's the project id: `prj_AndwsDMoec4m4144ikgE2ycmyTfW`

**Reply by swarnava** [reply]
Thanks! We cannot reproduce this unfortunately.

- Can you visit the deployment and share the deployment ID where this happen? `https://vercel.com/team-slug/project-slug/deployment-id`
- Can you also tell on which page the error occurs on as well as all the `response` headers from the server

**Reply by manu-r** [reply]
hi @swarnava 
this is the id from the url you pointed -> `AJiaFcPeW8Kc7e2GK9PAt7XTmAmA`

and the response from server
<img width="708" alt="Screenshot 2024-06-10 at 10 16 26 PM" src="https://github.com/vercel/community/assets/8032715/547f8bb2-f6dd-4263-afbb-c76c8becae7c">

EDIT: the error happens on `/gallery/*` path

**Reply by ztanner** [reply]
Hi @manu-r - thanks for sharing the details. When you have a moment, could you unpin the CLI and let us know if working as you expect in a fresh deploy?

**Reply by manu-r** [reply]
Hi @ztanner,

thanks for the quick fix. seems to be working now 👍

[a comment by an automated account is not shown]
